### DISCUSSION TITLE
How to modify response data

### DISCUSSION THREAD

**Discussion by tonxxd** [apollographql/apollo-server#7209] [Q&A] [answered]
I am trying to use the new v4 method to alter response before sending it out to the client but i am struggling to understand where is the response data in the context variable.
I am trying to parse ever
I am trying to do 
```js
async willSendResponse({ response }) {
 if (response.body.kind === "single" && "data" in response.body.singleResult  ) 
    console.log(response.body.singleResult.data)
}
```

But `response.body.singleResult.data` does not contain the raw response data

Any ideas?

**Comment by tonxxd** [comment, marked as answer]
I managed to solve by checking that response.body.singleResult.data contains other keys in addition to __schema.
Only at that point the `willSendResponse` response contains the final response.
```js
if (Object.keys(omit(response.body.singleResult.data, "__schema")).length > 0){
  console.log(omit(response.body.singleResult.data, "__schema"))
  // {"entities": [{"id":1}, ...]}
}
```

**Comment by glasser** [comment]
I'm not sure I understand what you're trying to do and what you're seeing.  It looks like you're trying to differentiate between introspection and non-introspection queries, based on your own answer?

**Comment by tonxxd** [comment]
I am trying to alter the response of the graphql api on the fly to automatically parse nested stringified json.
My data looks like
```graphql
Entitty {
  .. fields
  config: '{"field":"value","field2":"{\\"nested_stringifiedfield\\":3}"}'    // stringified json with nested stringified json
}
```
So i have to recursively explore the returned records to parse the deep nested data

**Reply by glasser** [reply]
OK, I'm not sure exactly why you weren't seeing anything under `data` then but sounds like you resolved your issue!

The transformations you are making will probably make your responses not standard GraphQL which will probably break some helpful tooling.
